File: frontend/lib/analytics.ts
import { apiClient, apiDownload } from "@/lib/api-client";
import type {
  AnalyticsFormat,
  AnalyticsPreset,
  CustomersAnalyticsResponse,
  DashboardResponse,
  OrdersAnalyticsResponse,
  ProductsAnalyticsResponse,
  RevenueAnalyticsResponse,
} from "@/types/analytics";

type RangeParams = {
  preset?: AnalyticsPreset;
  startDate?: string;
  endDate?: string;
};

function buildQuery(params: RangeParams = {}): string {
  const searchParams = new URLSearchParams();
  if (params.preset) {
    searchParams.set("preset", params.preset);
  }
  if (params.startDate) {
    searchParams.set("start_date", params.startDate);
  }
  if (params.endDate) {
    searchParams.set("end_date", params.endDate);
  }
  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export function buildAnalyticsExportUrl(
  format: AnalyticsFormat,
  params: RangeParams = {},
): string {
  const searchParams = new URLSearchParams();
  searchParams.set("format", format);
  if (params.preset) {
    searchParams.set("preset", params.preset);
  }
  if (params.startDate) {
    searchParams.set("start_date", params.startDate);
  }
  if (params.endDate) {
    searchParams.set("end_date", params.endDate);
  }
  const baseUrl = process.env.NEXT_PUBLIC_API_URL ?? "";
  const path = `/analytics/export?${searchParams.toString()}`;
  return baseUrl ? new URL(path, baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`).toString() : path;
}

export async function downloadAnalyticsExport(format: AnalyticsFormat, params: RangeParams = {}): Promise<void> {
  const searchParams = new URLSearchParams();
  searchParams.set("format", format);
  if (params.preset) {
    searchParams.set("preset", params.preset);
  }
  if (params.startDate) {
    searchParams.set("start_date", params.startDate);
  }
  if (params.endDate) {
    searchParams.set("end_date", params.endDate);
  }
  await apiDownload(`/analytics/export?${searchParams.toString()}`, `analytics-${format}`);
}

export async function getDashboardAnalytics(params: RangeParams = {}): Promise<DashboardResponse> {
  return apiClient<DashboardResponse>(`/dashboard${buildQuery(params)}`);
}

export async function getRevenueAnalytics(params: RangeParams = {}): Promise<RevenueAnalyticsResponse> {
  return apiClient<RevenueAnalyticsResponse>(`/analytics/revenue${buildQuery(params)}`);
}

export async function getOrdersAnalytics(params: RangeParams = {}): Promise<OrdersAnalyticsResponse> {
  return apiClient<OrdersAnalyticsResponse>(`/analytics/orders${buildQuery(params)}`);
}

export async function getProductsAnalytics(params: RangeParams = {}): Promise<ProductsAnalyticsResponse> {
  return apiClient<ProductsAnalyticsResponse>(`/analytics/products${buildQuery(params)}`);
}

export async function getCustomersAnalytics(params: RangeParams = {}): Promise<CustomersAnalyticsResponse> {
  return apiClient<CustomersAnalyticsResponse>(`/analytics/customers${buildQuery(params)}`);
}

File: backend/app/api/v1/analytics.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, Response
from sqlalchemy.orm import Session

from app.core.auth import Role
from app.core.config import Settings
from app.dependencies.auth import get_current_user, require_roles
from app.dependencies.deps import get_app_settings, get_db
from app.db.models.user import User
from app.schemas.analytics import (
    AnalyticsExportFormat,
    AnalyticsPreset,
    CustomersAnalyticsResponse,
    DashboardResponse,
    OrdersAnalyticsResponse,
    ProductsAnalyticsResponse,
    RevenueAnalyticsResponse,
)
from app.services.analytics_service import AnalyticsService

router = APIRouter(tags=["analytics"])


def _service(session: Session, settings: Settings) -> AnalyticsService:
    return AnalyticsService(session=session, settings=settings)


@router.get("/dashboard", response_model=DashboardResponse)
def get_dashboard(
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> DashboardResponse:
    return _service(session, settings).get_dashboard(
        current_user.company_id,
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/analytics/revenue", response_model=RevenueAnalyticsResponse, dependencies=[Depends(require_roles(Role.OWNER, Role.ADMIN, Role.MANAGER))])
def get_revenue_analytics(
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> RevenueAnalyticsResponse:
    return _service(session, settings).get_revenue_analytics(
        current_user.company_id,
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/analytics/orders", response_model=OrdersAnalyticsResponse, dependencies=[Depends(require_roles(Role.OWNER, Role.ADMIN, Role.MANAGER))])
def get_orders_analytics(
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> OrdersAnalyticsResponse:
    return _service(session, settings).get_orders_analytics(
        current_user.company_id,
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/analytics/products", response_model=ProductsAnalyticsResponse, dependencies=[Depends(require_roles(Role.OWNER, Role.ADMIN, Role.MANAGER))])
def get_products_analytics(
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> ProductsAnalyticsResponse:
    return _service(session, settings).get_products_analytics(
        current_user.company_id,
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/analytics/customers", response_model=CustomersAnalyticsResponse, dependencies=[Depends(require_roles(Role.OWNER, Role.ADMIN, Role.MANAGER))])
def get_customers_analytics(
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> CustomersAnalyticsResponse:
    return _service(session, settings).get_customers_analytics(
        current_user.company_id,
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/analytics/export", dependencies=[Depends(require_roles(Role.OWNER, Role.ADMIN, Role.MANAGER))])
def export_analytics(
    format: AnalyticsExportFormat = "csv",
    preset: AnalyticsPreset = "last_30_days",
    start_date: date | None = None,
    end_date: date | None = None,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db),
    settings: Settings = Depends(get_app_settings),
) -> Response:
    content, content_type, filename = _service(session, settings).export_analytics(
        current_user.company_id,
        export_format=format,  # type: ignore[arg-type]
        preset=preset,
        start_date=start_date,
        end_date=end_date,
    )
    return Response(
        content=content,
        media_type=content_type,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
